os/binfmt/binfmt_loadbinary.c

```diff
@@ -88,50 +88,76 @@ int load_binary(int binary_idx, FAR const char *filename, load_attr_t *load_attr
 	}
 
 #ifdef CONFIG_APP_BINARY_SEPARATION
-	/* Allocate the RAM partition to load the app into */
-	uint32_t *start_addr;
-	uint32_t size = load_attr->ram_size;
 	struct tcb_s *tcb;
-
-	if (mm_allocate_ram_partition(&start_addr, &size, load_attr->bin_name) < 0) {
-		berr("ERROR: Failed to allocate RAM partition\n");
-		errcode = ENOMEM;
-		goto errout;
-	}
 #endif
 
-	/* Allocate the load information */
-
-	bin = (FAR struct binary_s *)kmm_zalloc(sizeof(struct binary_s));
-	if (!bin) {
-		berr("ERROR: Failed to allocate binary_s\n");
-		errcode = ENOMEM;
-		goto err_free_partition;
-	}
-
-	/* Initialize the binary structure */
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD
+	bin = load_attr->binp;
+	/* If we find a non-null value for bin, it means that
+	 * we are in a reload scenario.
+	 */
+	if (bin) {
+		if (!bin->data_backup) {
+			errcode = -EINVAL;
+			berr("ERROR: Failed to find copy of data section from previous load\n");
+			goto errout_with_bin;
+		}
+
+		memcpy(bin->datastart, bin->data_backup, bin->datasize);
+		memset(bin->bssstart, 0, bin->bsssize);
+		bin->reload = false;
+	} else {
+#endif
 
-	bin->filename = filename;
-	bin->exports = NULL;
-	bin->nexports = 0;
-	bin->filelen = load_attr->bin_size;
-	bin->offset = load_attr->offset;
-	bin->stacksize = load_attr->stack_size;
-	bin->priority = load_attr->priority;
+		/* Allocate the load information */
+
+		bin = (FAR struct binary_s *)kmm_zalloc(sizeof(struct binary_s));
+		if (!bin) {
+			berr("ERROR: Failed to allocate binary_s\n");
+			errcode = ENOMEM;
+			goto errout_with_bin;
+		}
+
+		/* Initialize the binary structure */
+
+		bin->filename = filename;
+		bin->exports = NULL;
+		bin->nexports = 0;
+		bin->filelen = load_attr->bin_size;
+		bin->offset = load_attr->offset;
+		bin->stacksize = load_attr->stack_size;
+		bin->priority = load_attr->priority;
+		bin->compression_type = load_attr->compression_type;
 #ifdef CONFIG_APP_BINARY_SEPARATION
-	bin->uheap = (struct mm_heap_s *)start_addr;
-	bin->uheap_size = size;
+		bin->ramsize = load_attr->ram_size;
 #endif
-	bin->compression_type = load_attr->compression_type;
 
-	/* Load the module into memory */
+		/* Load the module into memory */
 
-	ret = load_module(bin);
-	if (ret < 0) {
-		errcode = -ret;
-		berr("ERROR: Failed to load program '%s': %d\n", filename, errcode);
-		goto errout_with_bin;
+		ret = load_module(bin);
+		if (ret < 0) {
+			errcode = -ret;
+			berr("ERROR: Failed to load program '%s': %d\n", filename, errcode);
+			goto errout_with_bin;
+		}
+
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD
+		if (!bin->data_backup) {
+			errcode = -ENOMEM;
```

Comment:
We didn't allocate the `data_backup` space, so this error code looks not good.

Reply:
Fixed.